src/field_finder.rs

```diff
@@ -9,134 +11,120 @@ pub enum TomlValue<'a> {
     ArrayOfTables(&'a mut ArrayOfTables),
 }
 
+#[derive(PartialEq, Eq)]
+pub enum DoInsert {
+    Yes,
+    No,
+}
+
 pub fn get_field<'a>(
     path: &[String],
     last_field: &String,
-    insert_if_not_exists: bool,
+    do_insert: DoInsert,
     doc: &'a mut Document,
-) -> Result<TomlValue<'a>, Error> {
+) -> Result<TomlValue<'a>> {
     let current_table = doc.as_table_mut();
 
-    descend_table(current_table, path, insert_if_not_exists, last_field)
+    descend_table(current_table, path, do_insert, last_field)
 }
 
 fn descend_table<'a>(
     table: &'a mut Table,
     path: &[String],
-    insert_if_not_exists: bool,
+    do_insert: DoInsert,
     last_field: &String,
-) -> Result<TomlValue<'a>, Error> {
+) -> Result<TomlValue<'a>> {
     let segment = match path.get(0) {
         Some(segment) => segment,
         None => return Ok(TomlValue::Table(table)),
     };
 
-    let val = if insert_if_not_exists {
-        // if next segment exists and is an integer insert array of tables
-        let insert_array_of_tables = match path.get(1) {
-            Some(segment) => segment.parse::<usize>().is_ok(),
-            None => last_field.parse::<usize>().is_ok(),
-        };
+    let val = match do_insert {
+        DoInsert::Yes => {
+            // if next segment exists and is an integer insert array of tables
+            let insert_array_of_tables = match path.get(1) {
+                Some(segment) => segment.parse::<usize>().is_ok(),
+                None => last_field.parse::<usize>().is_ok(),
+            };
 
-        let to_insert_as_backup = if insert_array_of_tables {
-            toml_edit::array()
-        } else {
-            toml_edit::table()
-        };
+            let to_insert_as_backup = if insert_array_of_tables {
+                toml_edit::array()
+            } else {
+                toml_edit::table()
+            };
 
-        table[segment].or_insert(to_insert_as_backup)
-    } else {
-        match table.get_mut(segment) {
-            Some(val) => val,
-            None => return Err(Error::new(ErrorKind::NotFound, "Path does not exist")),
+            table[segment].or_insert(to_insert_as_backup)
```

Comment:
It might be possible for this lookup to fail, `[]` syntax will panic if that happens, we might want to use `.get().conext("...")?` instead